cmd/object-handlers.go

```diff
@@ -1006,6 +1081,50 @@ func (api objectAPIHandlers) PutObjectPartHandler(w http.ResponseWriter, r *http
 		return
 	}
 
+	if objectAPI.IsEncryptionSupported() {
+		if IsSSECustomerRequest(r.Header) && !hasSuffix(object, slashSeparator) { // handle SSE-C requests
+			var key []byte
+			key, err = ParseSSECustomerRequest(r)
+			if err != nil {
+				writeErrorResponse(w, toAPIErrorCode(err), r.URL)
+				return
+			}
+
+			// Calculating object encryption key
+			var li ListPartsInfo
+			li, err = objectAPI.ListObjectParts(bucket, object, uploadID, 0, 1)
+			if err != nil {
+				writeErrorResponse(w, toAPIErrorCode(err), r.URL)
+				return
+			}
+
+			var objectEncryptionKey []byte
+			objectEncryptionKey, err = decryptObjectInfo(key, li.UserDefined)
+			if err != nil {
+				writeErrorResponse(w, toAPIErrorCode(err), r.URL)
+				return
+			}
+
+			sha := sha256.New() // derive part  encryption key
+			sha.Write(objectEncryptionKey)
+			sha.Write([]byte("-" + partIDString))
+			partEncryptionKey := sha.Sum(nil)
+
+			reader, err = sio.EncryptReader(reader, sio.Config{Key: partEncryptionKey})
+			if err != nil {
+				writeErrorResponse(w, toAPIErrorCode(err), r.URL)
+				return
+			}
+
+			info := ObjectInfo{Size: size}
+			hashReader, err = hash.NewReader(reader, info.EncryptedSize(), "", "") // do not try to verify encrypted content
```

Comment:
Why do not verify encrypted content here? If the request contains any checksum, I think it should be verified.

Reply:
we cannot since client didn't send encrypted @donatello

Reply:
@donatello

Reply:
@donatello The content sent by the client is plaintext and the object is encrypted on the server (SSE-C) - with a key derived from the client key and randomness. So by definition the client cannot sent a valid checksum for the encrypted content.